### PR TITLE
TextInfo move by codepoint characters function

### PR DESCRIPTION
**Pull request by mltony**
### Link to issue number:
This function is needed for both #8518 and #16050.
### Summary of the issue:
Suppose we have TextInfo that represents a paragraph of text:
```
> s = paragraphInfo.text
> s
'Hello, world!\r'
```
Suppose that we would like to put the cursor at the first letter of the word 'world'.
That means jumping to index 7:
```
> s[7:]
'world!\r'
```
The problem is that calling `paragraphInfo.move(UNIT_CHARACTER, 7, "start")` is not guaranteed to achieve desired effect. There are two main reasons for that:
1. In Wide character encoding, some 4-byte unicode characters are represented as two surrogate characters, whereas in pythonic string they would be represented by a single character.
2. In non-offset TextInfos (e.g. UIATextInfo) there is no guarantee on the fact that TextInfos.move(UNIT_CHARACTER, 1)would actually move by exactly 1 character. A good illustration of this is in Microsoft Word with UIA enabled always, the first character of a bullet list item would be represented by three pythonic characters:
    * Bullet character "•"
    * Tab character \t
    * And the first character of of list item per se.
3. The third problem of `TextInfo.move(UNIT_CHARACTER)` function is its performance in some implementations. In particular, moving by 10000 characters in Notepad++ takes over a second on a reasonably modern PC. I might not need to move by 10000 characters in my upcoming PRs, but I will need to move by a few thousands for sure since for sentence navigation I would need to move within a paragraph and some large paragraphs in typical texts can easily be few thousands characters. I need to find both beginning and end textInfos, and if each operation takes say 200ms, then we'd be wasting almost half a second on just moving by characters. Since there were previous concerns about sentence navigation being not fast enough, II would like to introduce this efficient implementation.

Here is how this can be done efficiently using this PR:
```
> info = paragraphInfo.moveToPythonicOffset(7)
> info.setEndPoint(paragraphInfo, "endToEnd")
> info.text
'world!\r'
```
### Description of user facing changes
N/A
### Description of development approach
1. For general case, I implemented binary-search-like algorithm. I explained it in great detail in the code. Please see `def moveToPythonicOffset` function in `textInfos\__init__.py`.
2. I provided optimized implementations for `OffsetsTextInfo` and `CompoundTextInfo`.
3. I refactored `textUtils.py` making it conformant to OOP style. I implemented `UTF8OffsetConverter` and dummy `IdentityOffsetConverter` as well as their abstract base class and a function `getOffsetConverter` that selects correct converter based on encoding. I renamed a couple of methods of `WideStringOffsetConverter` in order to remove the word `wide` - as now I would like to use similar methods for UTF8 converter, and it has nothing to do with wide strings.
### Testing strategy:
1. Unit tests
1. Tested in Notepad in presence of tricky 😂 characters (UIATextInfo).
2. Tested in Notepad++ in presence of 😂 characters (OffsetsTextInfo, UTF-8 encoding).
3. Tested in VSCode, in presence of 😂 characters (CompoundTextInfo, containing OffsetTextInfo with UTF-16 encoding).
4. Tested in Microsoft Word with bullet lists (UIATextInfo).
5. Tested in Chrome browse mode in presence of 😂 characters (OffsetsTextInfo,UTF-16  encoding).
6. Repeated above tests with Chinese, Arabic and Cyrillic characters.
7. Tested navigation by word using review cursor in browsers - this code path uses `WideStringOffsetConverter` - just to make sure my refactoring didn't break this class.
### Known issues with pull request:
N/A
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by mltony** [comment]
Also:
1. I haven't implemented unit tests yet. Given the complexity of binary search algorithm, I would want to implement them before merging. But I just want to see first if NVAccess is in general aligned with this approach before spending more time on this.
2. I anticipate  to be asked "why add a new function - why we can't add a new unit to existing `move` function". Let me try to address this proactively - here are a few reasons why I didn't go this way:
    * All use cases I have in mind would need to find a certain character within already existing paragraph textInfo. If we want to jump to a certain character of a paragraph, that implies we already have paragraph TextInfo and we're sure that the desired character lies somewhere inside. It would make sense to use this information since it's readily available instead of trying to perform move on an arbitrary and possibly collapsed textInfo.
    * As I mentioned before - performance is a concern. If we want to extend existing `find()` function, we'd have to have some sort of autoexpanding loop first, that first tries to figure out what would be enclosing textInfo that contains desired offset, and then call existing algorithm to actually find that offset. That seems wasteful to me and as I mentioned before, in any plausible use case we already have enclosing TextInfo available.

**Comment by Adriani90** [comment]
@mltony thanks for this. Very interesting. Can this be used also for character by character or word by word navigation? If yes, then this would probably open up a good way to solve also #11908, #2649, #13712, or #4431.

Also does this in your testing workwwell with compund characters in Asian or Arabic languages?

File: source/textUtils.py [hunk 1]
```diff
@@ -12,15 +12,102 @@
...
+class OffsetConverter:
+	decoded: str
+	
+	def __init__(self, text: str):
+		if not isinstance(text, str):
+			raise TypeError("Value must be of type str")
+		self.decoded: str = text
+
+	def __repr__(self):
+		return "{}({})".format(self.__class__.__name__, repr(self.decoded))
```
Comment (lukaszgo1) [inline source/textUtils.py:34]: Can't you use f-string here?

File: source/textUtils.py [hunk 2]
```diff
@@ -12,15 +12,102 @@
...
+		@param encodedEnd: The end offset in the wide character representation of the string.
+			This offset is exclusive.
+		@param raiseOnError: Raises an IndexError when one of the given offsets
+			exceeds L{encodedStringLength} or is lower than zero.
+			If C{False}, the out of range offset will be bounded to the range of the string.
+		@raise ValueError: if wideStringEnd < wideStringStart
+		"""
+		if encodedEnd is not None and encodedEnd < encodedStart:
+			raise ValueError(
+				"encodedEnd=%d must be greater than or equal to encodedStart=%d"
```
Comment (lukaszgo1) [inline source/textUtils.py:98]: Same as above W.R.t. usage of f-strings.

File: source/textUtils.py [hunk 3]
```diff
@@ -247,3 +305,126 @@ def isLowSurrogate(ch: str) -> bool:
...
+
+class UTF8OffsetConverter(OffsetConverter):
+	R"""
+	Object that holds a string in both its decoded and its UTF-8 encoded form.
+	The object allows for easy conversion between offsets in str type strings,
+	and offsets in UTF-8 encoded strings.
+
+	A single character in UTF-8 encoding might take 1, 2, or 4 bytes.
+	Examples of applications using UTF-8 encoding are all Scintilla-based text editors,
+	including Notepad++ and AkelPad.
```
Comment (lukaszgo1) [inline source/textUtils.py:318]: AkelPad is not Scintilla based and uses its own edit control. This comment should probably be clarified.

File: source/textUtils.py [hunk 4]
```diff
@@ -43,31 +130,20 @@ class WideStringOffsetConverter:
...
-			raise TypeError("Value must be of type str")
-		self.decoded: str = text
+		super().__init__(text)
 		self.encoded: bytes = text.encode(self._encoding, errors="surrogatepass")
 
-	def __repr__(self):
-		return "{}({})".format(self.__class__.__name__, repr(self.decoded))
-
 	@property
-	def wideStringLength(self) -> int:
```
Comment (lukaszgo1) [inline source/textUtils.py:56]: Removal of this property will be backward compatibility breaking. Some add-ons may rely on it (certainly one of mine does). Could you implement a back compat getter which logs the information that `wideStringLength` is deprecated and returns result of the new property? This should also be documented in the compatibility section in the change log after this is no longer a draft.

File: source/textInfos/__init__.py [hunk 5]
```diff
@@ -402,11 +403,11 @@ def _get_boundingRects(self):
 
 	def unitIndex(self,unit):
 		"""
-@param unit: a unit constant for which you want to retreave an index
-@type: string
-@returns: The 1-based index of this unit, out of all the units of this type in the object
-@rtype: int
-"""  
+		@param unit: a unit constant for which you want to retreave an index
+		@type: string
```
[the remark this answers is not in the document]
Comment (mltony) [reply source/textInfos/__init__.py:407]: That's legacy code - I am just fixing indentation.

File: source/textUtils.py (hunk 1 above)
Comment (mltony) [reply source/textUtils.py:34]: I'm just refactoring - that's legacy code.

File: source/textUtils.py (hunk 4 above)
Comment (mltony) [reply source/textUtils.py:56]: Added back old function names as aliases of new ones to preserve compatibility.

**Comment by LeonarddeR** [comment]
I really like this approach, but I'm not sure about the name `pythonic`.
According to the Python docs:
> Strings are immutable sequences of Unicode code points.

So may be code point offset is better.

**Comment by mltony** [comment]
@LeonarddeR,  Would be happy to rename it to something else more intuitive. But not sure that code point offset is more intuitive - I bet the term code point will definitely require people to look up the meaning of - whereas the term Pythonic is at least familiar to those devs who are aware of the difference between different offset schemes.

**Comment by mltony** [comment]
@seanbudd, could you take a look at this one?
This is blocking  #16050 and and my WIP PR for #8518.

File: source/textUtils.py (hunk 2 above)
Comment (seanbudd) [reply source/textUtils.py:96]: when refactoring legacy code, please modernise it. Otherwise it will never get fixed.

Comment (mltony) [reply source/textUtils.py]: Done.

**Comment by mltony** [comment]
@michaelDCurran, wondering if you can take a look at this PR? You gave green light for sentence navigation - this one is apparently required to make sentence navigation to work properly.

**Comment by michaelDCurran** [comment]
@mltony  I agree with @LeonarddeR  re the name. Pythonic is way too general... this is all "Python". Please rename it to moveToCodepointOffset.
I have looked at all code except textUtils.py so far. Looks pretty good to me. Please go ahead and start writing unit tests where you can.

**Comment by mltony** [comment]
@michaelDCurran, I addressed your comments, please have another look.
* Renamed to `moveToCodepointOffset`
* Added unit tests

**Comment by CyrilleB79** [comment]
@mltony could you also update the title and the initial description of this PR?
Pythonic ofsset -> codepoint offset
Thanks.

**Comment by mltony** [comment]
@michaelDCurran, @seanbudd, kindly ping - can either of you review this? This PR is blocking sentence navigation PR and style navigation PR. My leave is coming to an end soon and I really wanted to contribute these two PRs to NVDA before I go back to full time work.

**Comment by michaelDCurran** [comment]
This pr removes / renames public symbols in textUtils.py? Can you state how you have handled add-on compatibility? Assuming we can guarantee compatibility for with 2024.1, then I'm happy to approve.

**Comment by mltony** [comment]
In textUtils.py lines 236..238 I declare old function names and just assign new function names to them. Just like aliases. So any add-on calling these functions won't be affected. That only applies  to `WideStringOffsetConverter` class - since that's the only class in this file before this PR.

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**